File: src/test/java/io/bloco/faker/components/DateTest.java
package io.bloco.faker.components;

import io.bloco.faker.Faker;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import org.joda.time.DateTime;
import org.junit.Before;
import org.junit.Test;

import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.Matchers.anyOf;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.greaterThanOrEqualTo;
import static org.hamcrest.Matchers.lessThanOrEqualTo;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertThat;

public class DateTest {

    private Faker faker;
    private DateFormat dateFormat;

    @Before
    public void setUp() {
        faker = new Faker();
        dateFormat = new SimpleDateFormat("yyyy-MM-dd");
    }

    @Test
    public void between() throws Exception {
        String from = "1990-06-22";
        String to = "1990-06-23";

        java.util.Date fromDate = dateFormat.parse(from);
        java.util.Date toDate = dateFormat.parse(to);

        assertThat(faker.date.between(from, to), anyOf(equalTo(fromDate), equalTo(toDate)));
        assertThat(faker.date.between(fromDate, toDate), anyOf(equalTo(fromDate), equalTo(toDate)));
    }

    @Test
    public void forward() {
        java.util.Date today = new DateTime().withTimeAtStartOfDay().toDate();
        java.util.Date tomorrow = new DateTime().plusDays(1).withTimeAtStartOfDay().toDate();
        java.util.Date in100Days = new DateTime().plusDays(100).withTimeAtStartOfDay().toDate();

        assertThat(faker.date.forward(1),
                anyOf(equalTo(today), equalTo(tomorrow)));
        assertThat(faker.date.forward(100),
                allOf(greaterThanOrEqualTo(today), lessThanOrEqualTo(in100Days)));
    }

    @Test
    public void forwardDefault() {
        assertNotNull(faker.date.forward());
    }

    @Test
    public void backward() {
        java.util.Date today = new DateTime().withTimeAtStartOfDay().toDate();
        java.util.Date yesterday = new DateTime().minusDays(1).withTimeAtStartOfDay().toDate();
        java.util.Date minus100Days = new DateTime().minusDays(100).withTimeAtStartOfDay().toDate();

        assertThat(faker.date.backward(1),
                anyOf(equalTo(today), equalTo(yesterday)));
        assertThat(faker.date.backward(100),
                allOf(greaterThanOrEqualTo(minus100Days), lessThanOrEqualTo(today)));
    }

    @Test
    public void backwardDefault() {
        assertNotNull(faker.date.backward());
    }

    @Test
    public void birthday() {
        java.util.Date nextYear = new DateTime().plusYears(1).withTimeAtStartOfDay().toDate();
        java.util.Date in10Years = new DateTime().plusYears(10).withTimeAtStartOfDay().toDate();

        assertThat(faker.date.birthday(1, 10),
                allOf(greaterThanOrEqualTo(nextYear), lessThanOrEqualTo(in10Years)));
    }

    @Test
    public void birthdayDefault() {
        assertNotNull(faker.date.birthday());
    }
}

File: lib/src/test/kotlin/io/bloco/faker/components/DateTest.kt
package io.bloco.faker.components

import io.bloco.faker.Faker
import io.bloco.faker.helpers.toDate
import junit.framework.TestCase.assertNotNull
import junit.framework.TestCase.assertTrue
import org.junit.Test
import java.text.SimpleDateFormat
import java.time.LocalDate

class DateTest {
    private val faker = Faker()
    private val formatter = SimpleDateFormat("yyyy-MM-dd")

    @Test
    @Throws(Exception::class)
    fun between() {
        val from = "1990-06-22"
        val to = "1990-06-23"
        val fromDate = formatter.parse(from)
        val toDate = formatter.parse(to)

        assertTrue(faker.date.between(from, to) in fromDate..toDate)
        assertTrue(faker.date.between(fromDate, toDate) in fromDate..toDate)
    }

    @Test
    fun forward() {
        val today = LocalDate.now().toDate()
        val tomorrow = LocalDate.now().plusDays(1).toDate()
        val in100Days = LocalDate.now().plusDays(100).toDate()

        assertTrue(faker.date.forward(1) in today..tomorrow)
        assertTrue(faker.date.forward(100) in today..in100Days)
    }

    @Test
    fun forwardDefault() {
        assertNotNull(faker.date.forward())
    }

    @Test
    fun backward() {
        val today = LocalDate.now().toDate()
        val yesterday = LocalDate.now().minusDays(1).toDate()
        val minus100Days = LocalDate.now().minusDays(100).toDate()

        assertTrue(faker.date.backward(1) in yesterday..today)
        assertTrue(faker.date.backward(100) in minus100Days..today)
    }

    @Test
    fun backwardDefault() {
        assertNotNull(faker.date.backward())
    }

    @Test
    fun birthday() {
        val nextYear = LocalDate.now().plusYears(1).toDate()
        val in10Years = LocalDate.now().plusYears(10).toDate()
        assertTrue(faker.date.birthday(1, 10) in nextYear .. in10Years)
    }

    @Test
    fun birthdayDefault() {
        assertNotNull(faker.date.birthday())
    }
}